soundwire: add multi-lane support

bardliao opened the pull request:
The first 2 commits are from https://github.com/thesofproject/linux/pull/4704.

This PR adds SoundWire multi-lane support.

bardliao commented:
Hmm, the CI test Failure on TGLU_RVP_SDW-ipc4 is because that 2 1308 amps are on the same sdw link, and 1 controller data port maps to 2 peripheral data ports.  But I assume that controller data port and peripheral data port is 1:1 mapping. I need to think about how to handle this case.

plbossart commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -280,6 +280,35 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+ * mipi-sdw-lane-x-mapping properties, where x is 1..7, and the values for those
+ * properties are mipi-sdw-manager-lane-x or mipi-sdw-peripheral-link-y, where x
+ * is an integer between 1 to 7 if the lane is connected to a master lane, y is a
+ * character between A to E if the lane is connected to another slave lane.
```
probably best to use manager and peripheral in a consistent manner. The last sentence should probably be 

"
where x is an integer between 1 to 7 if the lane is connected to a Manager, y is a character between A to E if the lane is connected to another Peripheral.
"

bardliao replied:
Ack.

plbossart commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -280,6 +280,35 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+		len = strlen(prop_val);
+		/* The last character is enough to identify the connection */
+		prop->lane_maps[i] = prop_val[len - 1];
+		dev_err(dev, "[Chao] %c\n", prop->lane_maps[i]);
```
remove this dev_err or demote to dev_dbg.

plbossart commented on include/linux/soundwire/sdw.h:
```diff
@@ -363,6 +365,7 @@ struct sdw_dpn_prop {
  * @p15_behave: Slave behavior when the Master attempts a read to the Port15
  * alias
  * @lane_control_support: Slave supports lane control
+ * @lane_maps: Lane mapping for the slave
```
presumably only valid if lane_control_support is true?

bardliao replied:
> presumably only valid if lane_control_support is true?

Yes, and changed the description to "Lane mapping for the slave, only valid if lane_control_support is set"

plbossart commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -280,6 +280,35 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+		prop->lane_maps[i] = prop_val[len - 1];
+		dev_err(dev, "[Chao] %c\n", prop->lane_maps[i]);
+	}
+	return 0;
```
if the function returns 0 always, demote to void, or add a real error line 303?

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -314,7 +314,7 @@ static int sdw_select_row_col(struct sdw_bus *bus, int clk_freq)
 
 			if ((clk_freq - (frame_freq * SDW_FRAME_CTRL_BITS)) <
-			    bus->params.bandwidth)
+			    bandwidth)
```
I don't follow this patch at all.

First, we need to avoid that the frame size selected depends on the bandwidth required on each lane. It would be a disaster we had different results for each lane.

Second, this definition
````
include/linux/soundwire/sdw.h:#define SDW_FRAME_CTRL_BITS               48
````

is ONLY valid for lane0. There are no control bits on non-zero lanes, we can use the entire set of columns for audio data (or BPT/BRA).

bardliao replied:
Let me think more about it. My intent is to allocate lane according to available bandwidth on each lane. That's why I extend bandwidth to an array.

plbossart replied:
Right, but you could pick the lane which has the lowest bandwidth used, or conversely try to use the lane that's the most used. You have two dimensions to play with.

For corner cases, we will have to use max speed and max lanes. It's the intermediate cases that aren't so clear.

bardliao replied:
Hmm, I guess I can use a new lane if current bandwidth is not enough in sdw_compute_bus_params(). I will try it.

plbossart commented on drivers/soundwire/bus.h:
```diff
@@ -97,7 +97,7 @@ int sdw_find_col_index(int col);
 struct sdw_port_runtime {
 	int num;
-	int ch_mask;
+	u32 ch_mask;
```
commit title prefix: s/soundwore/soundwire/

plbossart commented on include/linux/soundwire/sdw.h:
```diff
@@ -950,10 +950,12 @@ void sdw_show_ping_status(struct sdw_bus *bus, bool sync_delay);
 struct sdw_port_config {
 	unsigned int num;
 	unsigned int ch_mask;
+	unsigned int lane_mask;
```
````lane 0 can only be used by port 1````

I don't think it's realistic to assume that some ports can only be used with non-zero lanes. By default I think the behavior is that all ports support Lane0. It'd be surprised if this wasn't the case, since that's the common denominator for all systems. It's possible that some non-zero lanes are not supported, e.g. a device could perfectly well support lane0..3 only.

````
The lane_maps[i] will be 0 by default and we can't tell if a lane is connected to lane 0 or not connected.
````

I couldn't follow the "lane is connected to lane0"....

bardliao replied:
> `lane 0 can only be used by port 1`

That is just an example to describe codec driver may want to specify a port to use a specific lane.
For example, headphone playback uses port 0 and lane 0. In that case, codec driver can set port_config.num=0 and port_config.lane_mask=0x1;

> 
> I don't think it's realistic to assume that some ports can only be used with non-zero lanes.

No, I don't assume that some ports can only be used with non-zero lanes.
The issue is that if peripheral lane x is not connected, the mipi-sdw-lane-x-mapping property will not be set and prop->lane_maps[x] will be 0 which is the same as if peripheral lane x is connected to manager lane 0.

> 
> ```
> The lane_maps[i] will be 0 by default and we can't tell if a lane is connected to lane 0 or not connected.
> ```
> 
> I couldn't follow the "lane is connected to lane0"....

Sorry, it should be "a peripheral lane is connected to manager lane 0".

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -2051,6 +2060,39 @@ int sdw_stream_add_slave(struct sdw_slave *slave,
+	for (i = 0; i < SDW_MAX_LANES; i++)
+		available_bandwidth[i] = bus->params.max_dr_freq - bus->params.bandwidth[i];
+
+	/* find available lane */
```
find first available lane?

I think the algorithm described tries to use a minimal number of lanes, instead of spreading the load on multiple lanes? It could be interesting if we try to change the frequency as well :-)

bardliao replied:
> I think the algorithm described tries to use a minimal number of lanes, instead of spreading the load on multiple lanes? It could be interesting if we try to change the frequency as well

Good point. I was thinking to use as less lanes as possible. Let me try to spread the load on multiple lanes.

plbossart replied:
I don't know what's best in terms of power consumption.

There are two options:
- optimize the clock and reduce the number of lanes, i.e. only use additional lanes when the bandwidth is too low at max speed
- always use additional lanes to reduce the speed.

If the speed is fixed, it's not clear either. On one hand there's a need to enable a new line, but with the NRZI encoding if there's nothing to transmit then there are no transitions on the data line.

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -1882,9 +1882,16 @@ int sdw_stream_add_master(struct sdw_bus *bus,
+	int lane[SDW_MAX_LANES];
+	int port_index = 0;
 	int ret;
+	int i;
```
"A controller should use the lane that connects to the peripheral"

That wording is controversial. 

First we are talking about a Manager. Controllers don't have a concept of Lanes, they are made of one or more Managers. Each Manager can have one or more Lanes. Different level.

Also, you could perfectly well have multiple lanes connecting a Manager and a Peripheral. For example to transmit 8ch 192kHz 24 bits we would need 4 lanes, one for each stereo pair.

bardliao replied:
The commit is to map the lane that connects to a specific peripheral lane. For example, peripheral lane 1 connects to manager lane 2. We already know which peripheral lane is used and need to get the manager lane by `slave_prop->lane_maps`.

plbossart reviewed (changes requested):
Good start @bardliao and @aiChaoSONG. I added a number of comments to make this PR more precise and clarify a number of technical opens.

bardliao committed:
soundwire: generic_bandwidth_allocation: select data lane

If a peripheral supports multi-lane, we can use data lane x to extend
the bandwidth. The patch suggests to select data lane x where x > 0
when bandwidth is not enough on data lane 0.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
soundwire: add lane_used_bandwidth in struct sdw_bus

To support multi-lane, we need to know how much bandwidth
is used on each lane. And to use the lane that has enough
bandwidth.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
soundwire: generic_bandwidth_allocation: add lane in sdw_group_params

All active streams with the same parameters are grouped together and the
params are store in the sdw_group struct. We compute the required
bandwidth for each group. However, each lane has individual bandwidth.
Therefore, we should separate different lanes in different params groups.
Add lane variable to separate params groups.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -280,6 +280,34 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+			continue;
+		len = strlen(prop_val);
+		/* The last character is enough to identify the connection */
+		prop->lane_maps[i] = prop_val[len - 1] - '0';
```
comment from @andy-shev still applies 
https://github.com/thesofproject/linux/pull/4857/files/856d130a14d412c016565ba7a429df38eb07b7a4..618319957b33c5a7638becc6ef7d3aefdeb0b822#r1529112951
"
IIRC, the string can be empty, this code then has a potential out-of-boundary access.
"

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -333,10 +333,18 @@ static int sdw_select_row_col(struct sdw_bus *bus, int clk_freq)
  */
 static int sdw_compute_bus_params(struct sdw_bus *bus)
 {
-	unsigned int curr_dr_freq = 0;
```
commit message s/bendwidth/bandwidth/

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +380,66 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+
+			/* Find a non-zero manager lane */
+			for (i = 1; i < SDW_MAX_LANES; i++) {
+				if (!slave_prop->lane_maps[i])
```
I think you should first check that all peripherals connected to that manager and using that stream do support the same lanes. Otherwise you may pick a lane that works for one of the peripherals but not for another.

This is a bit paranoid but it's not that crazy: we could have a big all-in-one device along with a companion amplifier on the same link. The big device will likely support multi-lane, the companion amplifier probably not.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +380,66 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+			}
+			list_for_each_entry(m_p_rt, &m_rt->port_list, port_node) {
+				m_p_rt->lane = m_lane;
+			}
```
that last loop should not be dependent on the number of peripherals connected. If there are N peripherals connected, this will be assigned N times, and possible one iteration overrides the previous settings.

It's really the same comment as above, the lane selection has to be the intersection of what all devices support, and the bandwidth counted once.

plbossart replied:
so in the new code update you verify the intersection but the bandwidth is still counted multiple times.

bardliao replied:
> so in the new code update you verify the intersection but the bandwidth is still counted multiple times.

Does the `bandwidth` here means `required_bandwidth`? It will be reset when we test each lane. So, it is not counted multiple times. `bus->params.bandwidth` is still used to select row and column. And if multi-lane is used, move the bandwidth from `bus->params.bandwidth` to bus->lane_used_bandwidth[i].

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -48,6 +50,9 @@ void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
 		list_for_each_entry(p_rt, &s_rt->port_list, port_node) {
+			if (p_rt->lane != t_data->lane)
+				continue;
+
```
not following this one, in which cases would the 'continue' happen?

bardliao replied:
When some p_rt uses lane 0 and the others uses lane 1. For example, headphone playback, headset capture, and speaker playback use lane 0 and DMIC capture uses lane 1.  We will have different param groups for different lanes. and _sdw_compute_port_params() will call sdw_compute_master_ports() -> sdw_compute_slave_ports() for each param group. And we only want to fill port params with the group param with the same lane.

plbossart replied:
I don't think it's possible that different streams share the same m_rt, and that means they would be treated separately.

The only case where I think a stream can be split across lanes is by channels, e.g. channel 0..3 on lane1 and channel 4..7 on lane2.

That makes me wonder if the t_data-> lane is really meaningful... I've lost how this is supposed to work now.

bardliao replied:
m_rt isn't shared by a stream, but sdw_group_params is. Please see below code in _sdw_compute_port_params().
```
	/* Run loop for all groups to compute transport parameters */
	for (i = 0; i < count; i++) {
		port_bo = 1;

		list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
			sdw_compute_master_ports(m_rt, &params[i], port_bo, hstop);

			port_bo += m_rt->ch_count * m_rt->stream->params.bps;
		}

		hstop = hstop - params[i].hwidth;
	}
```

Once two lanes are used, we will have 2 sdw_group_params. And `list_for_each_entry(m_rt, &bus->m_rt_list, bus_node)` will be called for each sdw_group_params. So, sdw_compute_master_ports() will be called twice for each m_rt, one for `&params[0]` , and the other for `&params[1]`. As p_rts in a m_rt could use different lanes, so, we can't just return like
```
        if (rate != params->rate)
                return;
```
And we need to pass `params->lane` to sdw_compute_slave_ports() as we only want to fill xport params when `p_rt->lane == params->lane`. Because the params like hstart, hstop is not for the p_rt if the params->lane doesn't match p_rt->lane.

plbossart replied:
@bardliao you really need to explain better in the commit message what groups are and why lanes should not share the groups.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -131,96 +138,126 @@ static void sdw_compute_master_ports(struct sdw_master_runtime *m_rt,
+	for (l = 0; l < SDW_MAX_LANES; l++) {
+		/* reset hstop for each lane */
+		hstop = bus->params.col - 1;
+		for (i = 0; i < count; i++) {
```
what does 'count' represent? I can't recall what this was used for.

bardliao replied:
'count' is the number of group params.

plbossart replied:
i guess I was never clear on what the 'group' represents...

bardliao replied:
I think the basic idea is to group the streams with the same rates together. Then calculate the required bandwidth etc. And the commit is to add lane into sdw_group_params{} because we should use different parameters for different lanes.

plbossart reviewed (commented):
Sorry for the late review @bardliao, see comments below.

plbossart commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -280,6 +280,39 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+		ret = kstrtou8(&prop_val[len - 1], 10, &lane);
+		if (ret)
+			return ret;
+		if (in_range(lane, 1, 7))
```
1, SDW_MAX_LANES - 1 ?

bardliao replied:
ack

plbossart commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -280,6 +280,39 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
+		ret = device_property_read_string(dev, prop_name, &prop_val);
+		if (ret)
+			continue;
+		len = strlen(prop_val);
```
check if len > 0

bardliao replied:
```
if (len < 1)
        return -EINVAL;
```
added

plbossart commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -382,6 +415,8 @@ int sdw_slave_read_prop(struct sdw_slave *slave)
 	sdw_slave_read_dpn(slave, prop->sink_dpn_prop, nval,
 			   prop->sink_ports, "sink");
 
+	sdw_slave_read_lane_mapping(slave);
```
do we want to discard errors? The error codes are not handled.

bardliao replied:
Changed to return sdw_slave_read_lane_mapping(slave);

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -370,16 +402,89 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+		/*
+		 * Find an available manager lane for the first Peripheral
+		 * Testing the first Peripheral is enough because we can't use multi-lane
+		 * if we can't find any available lane for the first Peripheral
```
I don't follow the logic.

If we can't find any available lanes for the first Peripheral, then we can't use multi-lane.
Buf if we want one available lane for the first Peripheral, we still need to check if other peripherals can use this lane.

Edit: and later you actually test all lanes so the comment is misleading. You *start* by testing the first peripheral and later verify that that this configuration works for all peripherals.

bardliao replied:
The logic here is to find an available lane from the first Peripheral, then check if the lane is available to all Peripherals.
The comment is to explain why I use `s_rt = list_first_entry(&m_rt->slave_rt_list, struct sdw_slave_runtime, m_rt_node);`

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -370,16 +402,89 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+					dev_dbg(bus->dev,
+						"some Peripherals are not connected to %d\n",
+						slave_prop->lane_maps[i]);
+					continue;
```
so here you continue to the next lane, but you've already increased required_bandwidth, so you will add it a second time.

bardliao replied:
> so here you continue to the next lane, but you've already increased required_bandwidth, so you will add it a second time.

No, `required_bandwidth` is set to 0 at the beginning of the loop. So when it continue, `required_bandwidth = 0;` will be called first, then increase `required_bandwidth`.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -370,16 +402,89 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+			}
+		}
+		/* Set Manager lanes */
+		/* Assume there is only one Manager in multi-lane configurations. */
```
I don't follow this comment, in the most agressive 192kHz 8ch bandwidth, we will have two links with two lanes each.

bardliao replied:
> I don't follow this comment, in the most agressive 192kHz 8ch bandwidth, we will have two links with two lanes each.

Hmm, I can't test the 2 links case for now. But for that case, sdw_compute_bus_params() will be called twice for each Manager, right?

bardliao replied:
Edited the comments. Hope it will be more readable.

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -1655,12 +1659,27 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
 
 		/* Compute params */
-		if (bus->compute_params) {
+		if (compute_params && bus->compute_params) {
```
The one question I have is whether all this works with the 192kHz 8ch case, where we will have to put 4 channels on one link and 4 channels on the other.

bardliao replied:
I think it should work, but I don't have an environment to test. I have a question, why don't we use 1 link with 4 lanes?

plbossart replied:
sorry @bardliao, I meant "The one question I have is whether all this works with the 192kHz 8ch case, where we will have to put 4 channels on one lane and 4 channels on a second lane."

bardliao replied:
> sorry @bardliao, I meant "The one question I have is whether all this works with the 192kHz 8ch case, where we will have to put 4 channels on one lane and 4 channels on a second lane."

Yes, I believe it will work.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -370,16 +402,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+			required_bandwidth = 0;
+			list_for_each_entry(m_p_rt, &m_rt->port_list, port_node) {
+				required_bandwidth += m_rt->stream->params.rate *
+					hweight32(m_p_rt->ch_mask) *
```
@bardliao where are the 'manager ports' ch_mask' defined?

It seems to be that the entire solution is based on a first split of the channels on different ports, but I don't know/understand where that first split is done.

Are we assuming that e.g. the manager code already looks at the bandwidth required in the dai hw_params and relies on multiple ports added with sdw_stream_add_master().

The other problem I have on the manager side is that we have a 1:1 mapping between PDI/port and DAI, so if we need to use multiple ports we would need multiple PDIs per dai.

Or it's possible that for now the solution moves entire streams to different lanes, but that's not solving the 8ch 192kHz problem where multi-lane is required.

bardliao replied:
> @bardliao where are the 'manager ports' ch_mask' defined?
> 
> It seems to be that the entire solution is based on a first split of the channels on different ports, but I don't know/understand where that first split is done.

Not exactly, the solution is not based on splitting channels to different ports. The solution is to check if current lane still has bandwidth for the new port, and use a different lane if there is no enough available bandwidth for the new port. So I don't split the channels to different port in this PR.

> 
> Are we assuming that e.g. the manager code already looks at the bandwidth required in the dai hw_params and relies on multiple ports added with sdw_stream_add_master().

Yes, if I expect that if a single stream needs more than 1 ports, it will be done in the dai hw_params. I am not sure if the manager code need to know and calculate the bandwidth accurately. In fact, I would expect that HDMI will use 2 PDIs/ports unconditionally.

> 
> The other problem I have on the manager side is that we have a 1:1 mapping between PDI/port and DAI, so if we need to use multiple ports we would need multiple PDIs per dai.

Yeah, PDI/port and DAI doesn't have to be 1:1 mapping.

> 
> Or it's possible that for now the solution moves entire streams to different lanes, but that's not solving the 8ch 192kHz problem where multi-lane is required.

Currently, we use a lane per stream. But that is because we use 1 port per stream. We will assign different lanes for different ports in the 8ch 192kHz case if we assign 2 ports for that stream.

plbossart replied:
Humm, I think we're doing this the wrong way, we should first support multiple ports per stream and then add multi-lane support.
The main reason is that your multi-lane parts loops over multiple ports, which isn't going to be tested fully *until* we do have multiple ports.

And the problem is that on paper the DAI-PDI does not need to be 1:1, but that's how it was used so far. I don't know how we would 'reserve' specific PDIs for specific DAIs...

bardliao replied:
> Humm, I think we're doing this the wrong way, we should first support multiple ports per stream and then add multi-lane support. The main reason is that your multi-lane parts loops over multiple ports, which isn't going to be tested fully _until_ we do have multiple ports.

We can test with multi-function codecs like rt722. We will need multi-lane when we run speaker playback, headset playback/capture, and dmic capture simultaneously with 4.8MHz clock rate.

> 
> And the problem is that on paper the DAI-PDI does not need to be 1:1, but that's how it was used so far. I don't know how we would 'reserve' specific PDIs for specific DAIs...

plbossart replied:
Yeah but the code has loops over multiple ports that will only be used with one port. That's why I think we absolutely need in addition to support multiple ports per dai, it's a miss today.

bardliao committed:
soundwire: mipi_disco: read lane mapping properties from ACPI

The DisCo for SoundWire 2.0 added support for the
'mipi-sdw-lane-<n>-mapping' property.

Co-developed-by: Chao Song <chao.song@linux.intel.com>
Signed-off-by: Chao Song <chao.song@linux.intel.com>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: rt722: get lane mapping property

Rt722 supports multi-lane and the driver doesn't call sdw_slave_read_prop()
to get all properties. Add sdw_slave_read_lane_mapping() to get the
required lane mapping property.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -493,6 +530,6 @@ int sdw_slave_read_prop(struct sdw_slave *slave)
 			   prop->sink_ports, "sink");
 
-	return 0;
+	return sdw_slave_read_lane_mapping(slave);
```
another possibility would be to export this function and have codec drivers call it directly, rather than read all properties just to read one that's required.

bardliao replied:
> another possibility would be to export this function and have codec drivers call it directly, rather than read all properties just to read one that's required.

Updated

plbossart commented on include/linux/soundwire/sdw.h:
```diff
@@ -874,6 +874,7 @@ struct sdw_master_ops {
  * meaningful if multi_link is set. If set to 1, hardware-based
  * synchronization will be used even if a stream only uses a single
  * SoundWire segment.
+<<<<<<< HEAD
```
merge issue here

bardliao replied:
oops

plbossart commented on include/linux/soundwire/sdw.h:
```diff
@@ -888,6 +889,7 @@ struct sdw_master_ops {
  * @multi_link: Store bus property that indicates if multi links
  * are supported. This flag is populated by drivers after reading
  * appropriate firmware (ACPI/DT).
+ * @lane_used_bandwidth: how much bandwidth is used by each lane
```
probably need to state the unit, I guess this is bits per second?

bardliao replied:
> probably need to state the unit, I guess this is bits per second?

Yes, it is m_rt->stream->params.rate * hweight32(m_p_rt->ch_mask) * m_rt->stream->params.bps.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -326,18 +326,50 @@ static int sdw_select_row_col(struct sdw_bus *bus, int clk_freq)
 	return -EINVAL;
 }
 
+static bool is_all_peripherals_connected(struct sdw_master_runtime *m_rt, unsigned int lane)
```
nit-pick: 'is' is singular, this sound weird in English.

Maybe 'check_all_peripherals_connected'

Also need to explain the purpose of this function. I think it's useful to determine if a stream can use Lane N in all peripherals used by the stream on a given link are connected on Lane N?

bardliao replied:
> nit-pick: 'is' is singular, this sound weird in English.
> 
> Maybe 'check_all_peripherals_connected'
> 
> Also need to explain the purpose of this function. I think it's useful to determine if a stream can use Lane N in all peripherals used by the stream on a given link are connected on Lane N?

Updated.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +404,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+		}
+		/*
+		 * Set Manager lanes. Configure the last m_rt in bus->m_rt_list only since
+		 * we don't wan't to touch other m_rts that are already working.
```
that's a bit controversial.

In theory the lane register could be changed for all streams so that a global optimization  of the bandwidth is handled.

by allocating a lane for just the last stream that become active, the algorithm looks for 'holes' in the bandwidth allocation. It's fine but it's subject to fragmentation and may fail at some point.

I think the patch should be explicit that given the relative similarity of the different streams used for now the suggested strategy works, but if we start mixing streams with completely different bandwidth requirements then this might need to be revisited.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +404,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+		}
+		dev_warn(bus->dev,
+			 "%s: could not find frame configuration for bus dr_freq %d try multi-lane\n",
+			 __func__, curr_dr_freq);
```
I don't think this should be a dev_warn since we know it's a legit configuration supported on some commercial devices.

bardliao replied:
> I don't think this should be a dev_warn since we know it's a legit configuration supported on some commercial devices.

Changed it to dev_dbg

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -133,94 +140,126 @@ static void sdw_compute_master_ports(struct sdw_master_runtime *m_rt,
-	for (i = 0; i < count; i++) {
-		port_bo = 1;
+	for (l = 0; l < SDW_MAX_LANES; l++) {
+		/* reset hstop for each lane */
```
now I am not following this comment since we said in the previous patches that only the last stream will use a different allocation, so not sure why all lanes need to use a HSTOP that's reset if we don't touch them.

bardliao replied:
> now I am not following this comment since we said in the previous patches that only the last stream will use a different allocation, so not sure why all lanes need to use a HSTOP that's reset if we don't touch them.

The hstop variable is not stored. We need to calculate the hstop for each port. And the hstop always starts with 'bus->params.col - 1' for each lane.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -284,7 +333,7 @@ static int sdw_compute_port_params(struct sdw_bus *bus)
 	/* Compute transport parameters for grouped streams */
 	ret = sdw_compute_group_params(bus, params,
-				       &group.rates[0], group.count);
+				       &group.rates[0], &group.lanes[0], group.count);
```
wondering if the order of patches should be modified to add first global changes to the allocation to make sure there's no regression by introducing all those modified loops in single-lane configuration, and then add the part where the multiple lanes are managed by RT722.

bardliao replied:
> wondering if the order of patches should be modified to add first global changes to the allocation to make sure there's no regression by introducing all those modified loops in single-lane configuration, and then add the part where the multiple lanes are managed by RT722.

Yes, done

bardliao replied:
> > wondering if the order of patches should be modified to add first global changes to the allocation to make sure there's no regression by introducing all those modified loops in single-lane configuration, and then add the part where the multiple lanes are managed by RT722.
> 
> Yes, done

Actually, no. `lane_maps[SDW_MAX_LANES];` is added in the `soundwire: mipi_disco: read lane mapping properties from ACPI` commit, and will be used in the follow-up commits. We can't move `soundwire: mipi_disco: read lane mapping properties from ACPI` to the back.

plbossart reviewed (changes requested):
some coments here and there to improve the patches, but request changes due to merge issue in a comment.

plbossart reviewed (dismissed):
sounds good to me for a first pass @bardliao !

@ujfalusi and @shumingfan please chime in.

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +410,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+		/*
+		 * Find available manager lanes that connected to the first Peripheral. No need
+		 * to check all Peripherals available lanes because we can't use multi-lane
+		 * if we can't find any available lane for the first Peripheral.
```
any particular reason the P is all peripherals is capitalized?

bardliao replied:
> any particular reason the P is all peripherals is capitalized?

Because Manager and Peripheral are terms in SDCA spec.

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +410,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+					m_rt->stream->params.bps;
+			}
+			if (required_bandwidth <= curr_dr_freq - bus->lane_used_bandwidth[i]) {
+				/* Check if m_lane is connected to all Peripherals */
```
what is m_lane? Doesnt seem to be define anywhere

bardliao replied:
m_lane indicates Manager lane. It is a local variable in sdw_compute_bus_params()

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +410,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+		s_rt = list_first_entry(&m_rt->slave_rt_list, struct sdw_slave_runtime, m_rt_node);
+		slave_prop = &s_rt->slave->prop;
+
+		for (i = 1; i < SDW_MAX_LANES; i++) {
```
this for {} is inside an if (i == clk_values). Do we want to use i again?

bardliao replied:
> this for {} is inside an if (i == clk_values). Do we want to use i again?

It should be fine, we will not test `if (i == clk_values)` later. But, yeah, it will be more readable to use another variable. Updated

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +410,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+				m_lane = slave_prop->lane_maps[i];
+				dev_dbg(&s_rt->slave->dev,
+					"M lane %d P lane %d can be used\n",
+					m_lane, i);
```
without much of a background on multi-lane, it is very difficult to understand what M lane and P lane is.

bardliao replied:
> without much of a background on multi-lane, it is very difficult to understand what M lane and P lane is.

A Manager can support multiple lanes, like lane 0, lane 1, and lane 2, lane 3. And a Peripheral also can have multiple lanes. Manager lane 0 and Peripheral lane 0 should be always connected. However, other lanes can be connected per request and no need to be 1:1 mapping. For example, Manager lane 1 can connect to Peripheral lane 2. So that we use M lane to represent Manager lane and P lane to Peripheral lane.

ranj063 commented on drivers/soundwire/stream.c:
```diff
@@ -1651,7 +1651,11 @@ EXPORT_SYMBOL(sdw_disable_stream);
 	struct sdw_master_runtime *m_rt;
+	struct sdw_port_runtime *p_rt;
+	bool compute_params = true;
+	unsigned int bandwidth;
```
can we keep the structs first?

ranj063 commented on drivers/soundwire/stream.c:
```diff
@@ -1665,12 +1669,27 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
+		bandwidth = m_rt->stream->params.rate * m_rt->ch_count * m_rt->stream->params.bps;
+		bus->params.bandwidth -= bandwidth - multi_lane_bandwidth;
+
+		/* No need to compute params since bus->params.bandwidth is not changed */
```
it would read better if you said "No need to compute params if bus->params.bandwidth is unchanged"

bardliao replied:
> it would read better if you said "No need to compute params if bus->params.bandwidth is unchanged"

Done

ranj063 commented on drivers/soundwire/stream.c:
```diff
@@ -1665,12 +1669,27 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
 
 		/* Compute params */
-		if (bus->compute_params) {
+		if (compute_params && bus->compute_params) {
```
why not use if (multi_lane_badnwidth > bandwidth && bus->comput_params) directly instead of adding the bool?

bardliao replied:
> why not use if (multi_lane_badnwidth > bandwidth && bus->comput_params) directly instead of adding the bool?

Updated to `if (multi_lane_badnwidth != bandwidth && bus->comput_params)`

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -372,16 +410,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+		}
+
+		if (!use_multi_lane) {
+			dev_err(bus->dev, "%s: could not find an available lane\n", __func__);
```
Probably better wording of the error is:
`"%s: multilane is not supported and no clock value found for bandwidth %d\n"` ?

bardliao replied:
> Probably better wording of the error is: `"%s: multilane is not supported and no clock value found for bandwidth %d\n"` ?

We may not find an available lane when multilane is supported. For example, a codec connect lane 0 and lane 1, but the lane bandwidth is currently used by other codecs.

ujfalusi replied:
Right, but I think the omission of the bandwidth target from the error print is not good, now the error can mean several things and from the log we will have no clue about the reason.

bardliao replied:
I got your point now. Updated to `"%s: multilane is not available and could not find clock value for bandwidth %d\n"`

ujfalusi reviewed (commented):
@bardliao, impressive set of patches! Only one small suggestion. Can be a followup fixup patch if you agree.

ranj063 commented on drivers/soundwire/mipi_disco.c:
```diff
@@ -366,6 +366,44 @@ static int sdw_slave_read_dpn(struct sdw_slave *slave,
 	return 0;
 }
 
+/*
```
@bardliao dont you need your sign off? Chao is no longer with Intel

ranj063 reviewed (dismissed):
looks good except for the first patch missing your sign-off @bardliao

bardliao commented:
> looks good except for the first patch missing your sign-off @bardliao

Updated to
```
Co-developed-by: Chao Song <chao.song@linux.intel.com>
Signed-off-by: Chao Song <chao.song@linux.intel.com>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>
```

bardliao commented:
@plbossart @ranj063 @ujfalusi I just found an issue when I tested with dynamic sdw bus clock. We need to clear `p_rt->ch_mask` when deprepare. Otherwise, sdw_compute_group_params() will still count the port bandwidth which is already stop when it is called in deprepare. And it will return error when we stop a stream and reduce the sdw clock. The same issue happens in the existing code, too. But we won't meet the issue because currently the sdw clock will never change.

bardliao commented:
> @plbossart @ranj063 @ujfalusi I just found an issue when I tested with dynamic sdw bus clock. We need to clear `p_rt->ch_mask` when deprepare. Otherwise, sdw_compute_group_params() will still count the port bandwidth which is already stop when it is called in deprepare. And it will return error when we stop a stream and reduce the sdw clock. The same issue happens in the existing code, too. But we won't meet the issue because currently the sdw clock will never change.

It will impact the multi-lane support. I need to think more about it.

ranj063 commented:
> > @plbossart @ranj063 @ujfalusi I just found an issue when I tested with dynamic sdw bus clock. We need to clear `p_rt->ch_mask` when deprepare. Otherwise, sdw_compute_group_params() will still count the port bandwidth which is already stop when it is called in deprepare. And it will return error when we stop a stream and reduce the sdw clock. The same issue happens in the existing code, too. But we won't meet the issue because currently the sdw clock will never change.
> 
> It will impact the multi-lane support. I need to think more about it.

@bardliao are you saying this PR is not ready for merging then? Also, what is the plan for testing multilane? I understand you have tested this locally but do we need to add tests to our CI also?

bardliao commented:
> > > @plbossart @ranj063 @ujfalusi I just found an issue when I tested with dynamic sdw bus clock. We need to clear `p_rt->ch_mask` when deprepare. Otherwise, sdw_compute_group_params() will still count the port bandwidth which is already stop when it is called in deprepare. And it will return error when we stop a stream and reduce the sdw clock. The same issue happens in the existing code, too. But we won't meet the issue because currently the sdw clock will never change.
> > 
> > 
> > It will impact the multi-lane support. I need to think more about it.
> 
> @bardliao are you saying this PR is not ready for merging then? Also, what is the plan for testing multilane? I understand you have tested this locally but do we need to add tests to our CI also?

After 1 and half day's investigation, I decide to roll back to the reviewed/approved and full tested version. I.e. remove the set p_rt->ch_mask = 0 in deprepare change. I will continue debug the issue when enabling dynamic sdw clock change. But in the meantime,  we can merge this if you agree. @ranj063 @plbossart @ujfalusi.

And yes, we need to setup a device to test multilane in CI. We will need a codec that can support multilane like rt722, rt712-vb

bardliao commented:
change: remove duplicated trace.

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -133,94 +140,126 @@ static void sdw_compute_master_ports(struct sdw_master_runtime *m_rt,
+				return -ENOMEM;
+
 			group->rates = rates;
+			group->lanes = lanes;
```
I think you need to move the `group->rates = rates;` further up (iow, add the lanes realloc after), before doing the krealloc() for the lanes. The group->rates is freed, but not NULL, so you will have double free error printed if lanes realloc fails.

bardliao replied:
done

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -133,94 +140,126 @@ static void sdw_compute_master_ports(struct sdw_master_runtime *m_rt,
-		column_needed += params[i].hwidth;
+			params[i].hwidth = (sel_col *
+				params[i].payload_bw + params[i].full_bw - 1) /
+				params[i].full_bw;
```
nitpick, but I think it would be easier to grasp this way:
```c
			params[i].hwidth = (sel_col * params[i].payload_bw +
					    params[i].full_bw - 1) / params[i].full_bw;
```

bardliao replied:
done

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -239,17 +279,26 @@ static int sdw_get_group_count(struct sdw_bus *bus,
 
+	group->lanes = kcalloc(group->max_size, sizeof(int), GFP_KERNEL);
+	if (!group->lanes)
+		return -ENOMEM;
```
you need to free `group->rates` - and set it to NULL to be safe.

bardliao replied:
> you need to free `group->rates` - and set it to NULL to be safe.

done
